### PR TITLE
Add ToOtlpLog unit tests for scopes (cont.)

### LINKED ISSUE #2873: OTLPLogExporter tests
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2871 - A very basic test is added here. Opening an issue to track adding more tests to OTLP Log Exporter. The exporter itself is not fully ready and is tracked here - https://github.com/open-telemetry/opentelemetry-dotnet/issues/2482

### PR DESCRIPTION
**Pull request by mfogliatto**
Fixes #2873

## Changes

- Add tests for multiple scope types
- Add tests for duplicated keys scenarios

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes N/A
* [X] Changes in public API reviewed (if applicable) N/A

### CODE REVIEW

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 1]
```diff
@@ -846,6 +847,244 @@ public void ToOtlpLog_WhenOptionsIncludeScopesIsTrue_ContainsScopeAttributeDoubl
...
+                new KeyValuePair<string, object>(scopeKey, "Some other scope value"),
+            }))
+            {
+                logger.LogInformation("Some log information message.");
+            }
+
+            // Assert.
+            var logRecord = logRecords.Single();
+            var otlpLogRecord = logRecord.ToOtlpLog(DefaultSdkLimitOptions);
+            var scopeAttributesCount = otlpLogRecord.Attributes.Count(_ => _.Key == scopeKey);
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:1013]: Add an assertion to ensure that the scope values are not overridden.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 2]
```diff
@@ -846,6 +847,244 @@ public void ToOtlpLog_WhenOptionsIncludeScopesIsTrue_ContainsScopeAttributeDoubl
...
+            Assert.Equal(2, scopeAttributesCount);
+        }
+
+        [Fact]
+        public void ToOtlpLog_WhenOptionsIncludeScopesIsTrue_AndMultpleScopeLevelsWithSameKeyAreAdded_ContainsAllAddedScopeValues()
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:1018]:
```suggestion
        public void ToOtlpLog_WhenOptionsIncludeScopesIsTrue_AndMultipleScopeLevelsWithSameKeyAreAdded_ContainsAllAddedScopeValues()
```

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 3]
```diff
@@ -846,6 +847,244 @@ public void ToOtlpLog_WhenOptionsIncludeScopesIsTrue_ContainsScopeAttributeDoubl
...
+            {
+                logger.LogInformation("Some log information message.");
+            }
+
+            // Assert.
+            var logRecord = logRecords.Single();
+            var otlpLogRecord = logRecord.ToOtlpLog(DefaultSdkLimitOptions);
+            var actualScope = TryGetAttribute(otlpLogRecord, scopeKey);
+            Assert.NotNull(actualScope);
+            Assert.Equal(scopeKey, actualScope.Key);
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:944]: Add an assertion for `otlpLogRecord.Attributes.Count` to ensure that both the keys (from log message and scope) are exported.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 4]
```diff
@@ -846,6 +847,244 @@ public void ToOtlpLog_WhenOptionsIncludeScopesIsTrue_ContainsScopeAttributeDoubl
...
+                    options.IncludeScopes = true;
+                    options.AddInMemoryExporter(logRecords);
+                });
+            });
+            var logger = loggerFactory.CreateLogger(nameof(OtlpLogExporterTests));
+
+            const string scopeKey = "Some scope key";
+
+            // Act.
+            using (logger.BeginScope(new List<KeyValuePair<string, object>> { new KeyValuePair<string, object>(scopeKey, "Some scope value") }))
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:1035]: You could probably update this test to a `Theory` which depending on a `bool` value could use either the same key or different keys for the scope levels.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 1 above)
Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: Thanks @utpilla ! I have updated the assertions to ensure scope values are not overriden and both remain in the log record.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 2 above)
Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: Thanks, fixed :)

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 3 above)
Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:950]: @utpilla thanks! I've just added the assertion to all tests where this is applicable.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 4 above)
Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: Thanks @utpilla . I'd prefer to keep the test setup as simple as possible. IMO adding an if/else clause would make the test less clear and introduce unnecessary complexity.

What if I pass the scopeKeys as inline data instead? This way we could still assert both scenarios.

Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: Just pushed the changes, please let me know your thoughts.

Comment (utpilla) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs]: Looks good.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 5]
```diff
@@ -840,12 +845,274 @@ public void ToOtlpLog_WhenOptionsIncludeScopesIsTrue_ContainsScopeAttributeDoubl
...
+                logger.LogInformation("Some log information message.");
+            }
+
+            // Assert.
+            var logRecord = logRecords.Single();
+            var otlpLogRecord = logRecord.ToOtlpLog(DefaultSdkLimitOptions);
+            Assert.Equal(2, otlpLogRecord.Attributes.Count);
+            var actualScope = TryGetAttribute(otlpLogRecord, scopeKey);
+            Assert.NotNull(actualScope);
+            Assert.Equal(scopeKey, actualScope.Key);
```
Comment (utpilla) [inline test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:950]: Add an assertion for scope value as well.

**Review by utpilla — APPROVED** [review]
Left some minor comments which are not blocking this PR.

Thanks @mfogliatto for adding these tests!

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 5 above)
Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:950]: Thanks @utpilla , will add in a follow up request.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs [hunk 6]
```diff
@@ -840,12 +845,274 @@ public void ToOtlpLog_WhenOptionsIncludeScopesIsTrue_ContainsScopeAttributeDoubl
...
+                new KeyValuePair<string, object>(scopeKey2, scopeValue2),
+            }))
+            {
+                logger.LogInformation("Some log information message.");
+            }
+
+            // Assert.
+            var logRecord = logRecords.Single();
+            var otlpLogRecord = logRecord.ToOtlpLog(DefaultSdkLimitOptions);
+            var allScopeValues = otlpLogRecord.Attributes
```
[the remark this answers is not in the document]
Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:1023]: Thanks! Added in a follow up PR.

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 5 above)
Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:950]: Added in https://github.com/open-telemetry/opentelemetry-dotnet/pull/4610

File: test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mfogliatto) [reply test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs:1023]: Added in https://github.com/open-telemetry/opentelemetry-dotnet/pull/4610

**Pull request merged**